verify-credentials.js

```diff
@@ -0,0 +1,84 @@
+ * deleting the user when we get one of these codes, store a 'deactivatedAt'
+ * timestamp on the user object. Users with a non-null deactivatedAt
+ * get their credentials retried once per day for 30 days, after which (TOD)
+ * they should be deleted. Regular operations like checking blocks or
+ * streaming are not performed for users with non-null deactivatedAt.
+ */
+function verifyCredentials(user) {
+  twitter.account('verify_credentials', {}, user.access_token,
+    user.access_token_secret, function(err, results) {
+      if (err && err.data) {
+        // For some reason the error data is given as a string, so we have to
+        // parse it.
+        var errJson = JSON.parse(err.data);
+        if (errJson.errors &&
+            errJson.errors.some(function(e) { return e.code === 89 })) {
+          logger.warn('User', user, 'revoked app.');
+          user.deactivatedAt = new Date();
+        } else if (err.statusCode === 404) {
+          //Testing for issue #146 indicates calling verify_credentials
+          //for a suspended account will return success, not a 404. Leaving
+          //this code in-place for robustness and in case there are
+          //differences between suspended accounts we may not be aware of.
+          logger.warn('User', user, 'deactivated or suspended.')
+          user.deactivatedAt = new Date();
+        } else {
+          logger.warn('User', user, 'verify_credentials', err.statusCode);
+        }
+      } else {
+        //We need a second lookup to the users.show.json endpoint to detect
+        //suspension status for issue #146
+        twitter.users('show',
+          { user_id: user.uid },
+          user.access_token,
+          user.access_token_secret,
+          function(err, response) {
+            if (err) {
+              // Testing using a suspended account does not seem to return
+              // 403 or 404. Leaving this here for robustness, see comments
+              // in error case of verify_credentials response handler.
+              if (err.statusCode === 403 || err.statusCode === 404) {
+                logger.warn('User', user, 'is deactivated or suspended. Marking so', err.statusCode);
+                user.deactivatedAt = new Date();
+              } else {
+                logger.warn('User', user, 'unknown error on /user/lookup', err);
+              }
+            } else {
+              // Testing for #146 showed the response.suspended field to be
+              // a reliable indicator of whether an account was suspended.
+              if (response && response.suspended === true)
+              {
+                logger.warn('User', user, 'is suspended. Marking so');
+                user.deactivatedAt = new Date();
+              } else {
+                logger.info('User', user, 'has not revoked app or deactivated.');
+                user.deactivatedAt = null;
+              }
+            }
+          }
+        );
+      }
+      if (user.changed()) {
```

Comment:
This is where Node's async model gets really headachey: Since one of the paths that could update the value is in the `/users/show.json` callback, the `user.save()` call will fire before the callback is called.

Quick fix: Move the existing user.save() call inside first `if` clause, and put a copy of the user.save() call inside the `/users/show.json` callback.

Nicer fix: Promises actually help a lot here. They let you easily make both Twitter requests in parallel, wait for both to be done, and then handle both responses together synchronously.

```
var verifyPromise = Q.ninvoke(twitter, 'account', 'verify_credentials' ...)
var usersShowPromise = Q.ninvoke(twitter, 'users', 'show', ...)

Q.spread([verifyPromise, usersShowPromise]
  function(verifyResponse, usersShowResponse) {
  if (verifyResponse...) {
  } else if (usersShowResponse.suspended) {
    ...
  }
}
```

Anyway, I am happy with either solution.